src/util/file_watch.c

```diff
@@ -0,0 +1,308 @@
+                                                poll_watched_file,
+                                                fw_ctx);
+    if (!fw_ctx->poll_check.timer) {
+        return EIO;
+    }
+
+    return EOK;
+}
+
+static void poll_watched_file(struct tevent_context *ev,
+                              struct tevent_timer *te,
+                              struct timeval t, void *ptr)
+{
+    int ret, err;
+    struct stat file_stat;
+    struct file_watch_ctx *fw_ctx;
+
+    fw_ctx = talloc_get_type(ptr, struct file_watch_ctx);
+
+    ret = stat(fw_ctx->filename, &file_stat);
+    if (ret < 0) {
+        err = errno;
+        DEBUG(SSSDBG_IMPORTANT_INFO,
+              "Could not stat file [%s]. Error [%d:%s]\n",
+              fw_ctx->filename, err, strerror(err));
+        return;
+    }
+
+    if (file_stat.st_mtime != fw_ctx->poll_check.modified) {
+        /* Parse the file and invoke the callback */
+        /* Note: this will fire if the modification time changes into the past
+         * as well as the future.
+         */
+        DEBUG(SSSDBG_TRACE_INTERNAL, "File [%s] changed\n", fw_ctx->filename);
+        fw_ctx->poll_check.modified = file_stat.st_mtime;
+
+        /* Tell the caller the file changed */
+        fw_ctx->cb(fw_ctx->filename, fw_ctx->cb_arg);
+    }
+
+    ret = create_poll_timer(fw_ctx);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_IMPORTANT_INFO,
+              "Error: File [%s] no longer monitored for changes!\n",
+              fw_ctx->filename);
+    }
+}
+
+
+static int watched_file_inotify_cb(const char *filename,
+                                  uint32_t flags,
+                                  void *pvt)
+{
+    struct file_watch_ctx *fw_ctx;
+
+    DEBUG(SSSDBG_TRACE_LIBS,
+          "Received inotify notification for %s\n", filename);
+
+    fw_ctx = talloc_get_type(pvt, struct file_watch_ctx);
+    if (fw_ctx == NULL) {
+        return EINVAL;
+    }
+
+    if (strcmp(fw_ctx->filename, filename) == 0) {
+        fw_ctx->cb(fw_ctx->filename, fw_ctx->cb_arg);
+    }
+
+    return EOK;
+}
+
+
+static int try_inotify(struct file_watch_ctx *fw_ctx)
+{
+#ifdef HAVE_INOTIFY
+    struct snotify_ctx *snctx;
+    /* We will queue the file for update in one second.
+     * This way, if there is a script writing to the file
+     * repeatedly, we won't be attempting to update multiple
+     * times.
+     */
+    struct timeval delay = { .tv_sec = 1, .tv_usec = 0 };
+
+    snctx = snotify_create(fw_ctx, fw_ctx->ev, SNOTIFY_WATCH_DIR,
+                           fw_ctx->filename, &delay,
+                           IN_DELETE_SELF | IN_CLOSE_WRITE | IN_MOVE_SELF | \
+                           IN_CREATE | IN_MOVED_TO | IN_IGNORED,
+                           watched_file_inotify_cb, fw_ctx);
+    if (snctx == NULL) {
+        return EIO;
+    }
+
+    return EOK;
+#else
+    return EINVAL;
+#endif /* HAVE_INOTIFY */
+}
+
+static errno_t fw_watch_file_fallback(TALLOC_CTX *parent_ctx,
```

Comment:
if you agree, and assuming the fallback will always be polling, `fw_watch_file_poll()` could be a more descriptive function name

Reply:
Good idea.